### PR TITLE
InitializeParameter refactorings support for local & anonymous functions

### LINKED ISSUE #20983: No suggestion to add null check for inner function
Ported from https://devdiv.visualstudio.com/DevDiv/_workitems/edit/466136

Paste: static void Foo(Stuff outer) { int Bar(Stuff inner) { return 0; } Bar(outer); } Click on `outer`. There is an offer to add a null check. Click on `inner`. There is no such offer.

### PR DESCRIPTION
**Pull request by Neme12**
fixes #20983

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ArrowExpressionClauseSyntaxExtensions.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ internal static class ArrowExpressionClauseSyntaxExtensions
             return true;
         }
 
-        private static StatementSyntax ConvertToStatement(
+        public static StatementSyntax ConvertToStatement(
```
Comment (Neme12) [inline src/Workspaces/CSharp/Portable/Extensions/ArrowExpressionClauseSyntaxExtensions.cs]: This is obviously not ideal... The thing is that this method is also useful for converting expression lambdas to statements, not just expression-bodied members. Maybe it could be moved to ExpressionSyntaxExtensions and then there would be a separate class for lambda extensions? I didn't want to restructure the code significantly in my first contribution, especially in layers below that I wasn't supposed to touch, just to make a simple change in a refactoring.

**Comment by Neme12** [comment]
@CyrusNajmabadi could you please give me some feedback? there's a few rough spots I'm not sure what to do about

File: src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb [hunk 2]
```diff
@@ -268,19 +268,118 @@ end class")
...
+                end sub
+    end sub
+end class",
+"
+Imports System
+
+class C
+    public sub new()
+        dim f = sub (s as string)
+
```
Comment (Neme12) [inline src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: All these blank lines in VB... the VB syntax generator does that for some reason, I don't know why or what the purpose of that is so I'm not changing it. It's in `VisualBasicSyntaxGenerator.WithStatements` but if you look at `CSharpSyntaxGenerator.WithStatements`, there's no code like that.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: yeah... that's not good.   we'd def want to fix this.  or at least open a bug here.  we don't do this for normal sub/function methods right?

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 3]
```diff
@@ -11,8 +11,42 @@ namespace Microsoft.CodeAnalysis.CSharp.InitializeParameter
...
+            switch (functionDeclaration)
+            {
+                case BaseMethodDeclarationSyntax methodDeclaration:
+                    return (SyntaxNode)methodDeclaration.Body ?? methodDeclaration.ExpressionBody;
+                case LocalFunctionStatementSyntax localFunction:
+                    return (SyntaxNode)localFunction.Body ?? localFunction.ExpressionBody;
+                case AnonymousFunctionExpressionSyntax anonymousFunction:
+                    return (SyntaxNode)anonymousFunction.Body;
+                default:
+                    Debug.Fail($"Unexpected {nameof(functionDeclaration)} type");
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: Throw.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 4]
```diff
@@ -11,8 +11,42 @@ namespace Microsoft.CodeAnalysis.CSharp.InitializeParameter
...
+            switch (functionDeclaration)
+            {
+                case BaseMethodDeclarationSyntax methodDeclaration:
+                    return methodDeclaration.SemicolonToken;
+                case LocalFunctionStatementSyntax localFunction:
+                    return localFunction.SemicolonToken;
+                case AnonymousFunctionExpressionSyntax _:
+                    return null;
+                default:
+                    Debug.Fail($"Unexpected {nameof(functionDeclaration)} type");
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: throw.

File: src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb (hunk 2 above)
Comment (Neme12) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: yes this only happens with lambdas

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 5]
```diff
@@ -24,33 +58,30 @@ public static SyntaxNode TryGetLastStatement(IBlockOperation blockStatementOpt)
...
-                else
-                {
-                    editor.SetStatements(methodDeclaration, ImmutableArray.Create(declaration, statement));
-                }
+                //Add the new statement as the first/last statement of the new block 
+                // depending if we were asked to go after something or not.
+                editor.SetStatements(functionDeclaration, statementToAddAfterOpt == null
+                    ? ImmutableArray.Create(statement, convertedStatement)
+                    : ImmutableArray.Create(convertedStatement, statement)
+                    );
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: put on previous line.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 6]
```diff
@@ -73,27 +104,33 @@ public static SyntaxNode TryGetLastStatement(IBlockOperation blockStatementOpt)
...
+                editor.SetStatements(functionDeclaration, ImmutableArray.Create(statement));
             }
         }
 
-        private static bool CreateReturnStatement(BaseMethodDeclarationSyntax declarationSyntax)
+        public static bool IsExpressionBody(SyntaxNode body)
         {
-            switch(declarationSyntax)
+            // either expression lambda or expression bodied member
+            return body is ExpressionSyntax || body is ArrowExpressionClauseSyntax;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: consider using => here.

File: src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb (hunk 2 above)
Comment (Neme12) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: so i checked again and it does create blank lines between the statements in methods, but not before the first one... in lambdas it also creates a blank line before the first statement

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -61,44 +59,43 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
...
             var parameter = (IParameterSymbol)semanticModel.GetDeclaredSymbol(parameterNode, cancellationToken);
-            if (!method.Parameters.Contains(parameter))
+            if (parameter == null || parameter.Name == "")
             {
                 return;
             }
 
-            if (parameter.Name == "")
+            var method = (IMethodSymbol)parameter.ContainingSymbol;
+            if (method == null ||
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: null check not needed.

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -61,44 +59,43 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
...
+                method.IsAbstract ||
+                method.IsExtern ||
+                method.PartialImplementationPart != null ||
+                method.ContainingType?.TypeKind == TypeKind.Interface)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: ?. not needed.

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs [hunk 9]
```diff
@@ -107,16 +104,12 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
             if (bodyOpt != null)
             {
                 blockStatementOpt = semanticModel.GetOperation(bodyOpt, cancellationToken) as IBlockOperation;
-                if (blockStatementOpt == null)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: this is just paranoia until IOperation is complete.

File: src/Features/VisualBasic/Portable/InitializeParameter/InitializeParameterHelpers.vb [hunk 10]
```diff
@@ -1,32 +1,59 @@
...
+                Dim multiLineLambda = DirectCast(functionDeclaration, MultiLineLambdaExpressionSyntax)
+                return multiLineLambda.Statements
+            ElseIf TypeOf functionDeclaration Is SingleLineLambdaExpressionSyntax Then
+                Dim singleLineLambda = DirectCast(functionDeclaration, SingleLineLambdaExpressionSyntax)
+                Dim convertedStatement = If(TypeOf singleLineLambda.Body Is StatementSyntax,
+                    DirectCast(singleLineLambda.Body, StatementSyntax),
+                    SyntaxFactory.ReturnStatement(DirectCast(singleLineLambda.Body, ExpressionSyntax)))
+                return SyntaxFactory.List(ImmutableArray.Create(convertedStatement))
+            Else
+                Debug.Fail($"Unexpected {NameOf(functionDeclaration)} type")
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/InitializeParameter/InitializeParameterHelpers.vb]: throw.

File: src/Workspaces/CSharp/Portable/Extensions/ArrowExpressionClauseSyntaxExtensions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ArrowExpressionClauseSyntaxExtensions.cs]: i would leave it here for now.

**Comment by CyrusNajmabadi** [comment]
Looking awesome!

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 5 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: just the `);`?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: yup

File: src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb (hunk 2 above)
Comment (Neme12) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: I was suspicious this might have been caused by the call to Isolate (which calls PreserveTrivia) in VisualBasicSyntaxGenerator.WithStatements - there's no such thing in CSharpSyntaxGenerator, but that's not the case, I get same behavior if I directly call WithStatementsInternal. It must be some sort of formatting issue.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 3 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: fixed

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 4 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: fixed

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 5 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: fixed

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 6 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: fixed

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs (hunk 7 above)
Comment (Neme12) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: fixed

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs (hunk 8 above)
Comment (Neme12) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: fixed

File: src/Features/VisualBasic/Portable/InitializeParameter/InitializeParameterHelpers.vb (hunk 10 above)
Comment (Neme12) [reply src/Features/VisualBasic/Portable/InitializeParameter/InitializeParameterHelpers.vb]: fixed

**Comment by Neme12** [comment]
I found a bug: in VB lambdas the null checks don't get ordered correctly and are offered even if there is an existing one. The IOperation wireup must be wrong there.

**Comment by Neme12** [comment]
Fixed.

File: src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb (hunk 2 above)
Comment (Neme12) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: @CyrusNajmabadi I identified the issue. It's in [ElasticTriviaFormattingRule.TopLevelStatement](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Workspaces/Formatting/Rules/ElasticTriviaFormattingRule.vb,370). It's a one-line fix to add `TypeOf statement Is LambdaHeaderSyntax`. Should I include it in this PR?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasicTest/InitializeParameter/AddParameterCheckTests.vb]: I'd have no problem with that.

**Comment by Neme12** [comment]
All fixed... I think this is ready

File: src/EditorFeatures/VisualBasicTest/CodeActions/IntroduceVariable/IntroduceVariableTests.vb [hunk 11]
```diff
@@ -1282,7 +1282,6 @@ End Module",
 Module Program
     Sub Main
         Dim a = Sub(x As Integer)
-
                     If True Then
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/CodeActions/IntroduceVariable/IntroduceVariableTests.vb]: these deletions are good.  this PR also fixes a VB formatting bug around elastic trivia and lambdas.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs [hunk 12]
```diff
@@ -15,31 +16,33 @@ namespace Microsoft.CodeAnalysis.CSharp.InitializeParameter
...
+            if (InitializeParameterHelpers.IsExpressionBody(body))
             {
-                return arrowExpressionClauseSyntax.TryConvertToStatement(
-                    semicolonToken: SyntaxFactory.Token(SyntaxKind.SemicolonToken), 
-                    createReturnStatementForExpression: false, 
-                    statement: out var _);              
+                return InitializeParameterHelpers.TryConvertExpressionBodyToStatement(body,
+                    semicolonToken: SyntaxFactory.Token(SyntaxKind.SemicolonToken),
+                    createReturnStatementForExpression: false,
+                    statement: out var _);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: do we have a test that exercises this codepath when TryConvertExpressionBodyToStatement returns false?

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromParameterCodeRefactoringProvider.cs [hunk 13]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Composition;
+using System.Threading;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromParameterCodeRefactoringProvider.cs]: why?

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 14]
```diff
@@ -2,17 +2,63 @@
...
+        public static bool IsFunctionDeclaration(SyntaxNode node)
+            => node is BaseMethodDeclarationSyntax
+            || node is LocalFunctionStatementSyntax
+            || node is AnonymousFunctionExpressionSyntax;
+
+        public static IBlockOperation GetBlockOperation(SyntaxNode functionDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            var bodyOpt = GetBody(functionDeclaration);
+            if (bodyOpt == null)
+                return null;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: curlies always required.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 15]
```diff
@@ -2,17 +2,63 @@
...
+        private static SyntaxNode GetBody(SyntaxNode functionDeclaration)
+        {
+            switch (functionDeclaration)
+            {
+                case BaseMethodDeclarationSyntax methodDeclaration:
+                    return (SyntaxNode)methodDeclaration.Body ?? methodDeclaration.ExpressionBody;
+                case LocalFunctionStatementSyntax localFunction:
+                    return (SyntaxNode)localFunction.Body ?? localFunction.ExpressionBody;
+                case AnonymousFunctionExpressionSyntax anonymousFunction:
+                    return (SyntaxNode)anonymousFunction.Body;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: this cast should not be necessary.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 14 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: :( ok

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 16]
```diff
@@ -2,17 +2,63 @@
...
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(functionDeclaration);
+            }
+        }
+
+        private static SyntaxToken? GetSemicolonToken(SyntaxNode functionDeclaration)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: preference is to call this TryGetSemicolonToken to indicate that null is a reasonable value you can get back here.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 17]
```diff
@@ -24,33 +70,29 @@ public static SyntaxNode TryGetLastStatement(IBlockOperation blockStatementOpt)
...
+            if (IsExpressionBody(body))
             {
-                // If this is a => method, then we'll have to convert the method to have a block
-                // body.  Add the new statement as the first/last statement of the new block 
-                // depending if we were asked to go after something or not.
-                if (!methodDeclaration.ExpressionBody.TryConvertToStatement(methodDeclaration.SemicolonToken, CreateReturnStatement(methodDeclaration), out var declaration))
-                {
+                var semicolonToken = GetSemicolonToken(functionDeclaration) ?? SyntaxFactory.Token(SyntaxKind.SemicolonToken);
+
+                if (!TryConvertExpressionBodyToStatement(body, semicolonToken, !method.ReturnsVoid, out var convertedStatement))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: curlies always required.

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs [hunk 18]
```diff
@@ -61,62 +59,49 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
...
+            var functionDeclaration = parameterNode.FirstAncestorOrSelf<SyntaxNode>(IsFunctionDeclaration);
+            if (functionDeclaration == null)
             {
                 return;
             }
 
             var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
             var parameterDefault = syntaxFacts.GetDefaultOfParameter(parameterNode);
-
+
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: unnecessary whitespace.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 15 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: it's not, i wanted it to look the same as on previous lines... i'll remove it

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs [hunk 19]
```diff
@@ -61,62 +59,49 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
...
-            if (bodyOpt != null)
-            {
-                blockStatementOpt = semanticModel.GetOperation(bodyOpt, cancellationToken) as IBlockOperation;
-                if (blockStatementOpt == null)
-                {
-                    return;
-                }
-            }
-
+            var blockStatementOpt = GetBlockOperation(functionDeclaration, semanticModel, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: i would still like the null check for this.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide for another review.

**Comment by CyrusNajmabadi** [comment]
@sharwell too for community PR.

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs (hunk 19 above)
Comment (Neme12) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: do what if it's null? it can be null as mentioned in the comment

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: ah, i see.  ok then :)

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 16 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: isn't that what the `?` is for?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: If you look at the signature yes.  But at the reference point it's not clear.  We find the code reads more clearly in certain cases if you write it as "TryXXX".  (And yes, this isn't consistently applied).

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs (hunk 12 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: yes: https://github.com/Neme12/roslyn/blob/957882b0ce32707c6540872dbebb51c4cf5eb03e/src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs#L1170

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: great.  thanks!

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs (hunk 18 above)
Comment (Neme12) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs]: removed

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 17 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: done

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 16 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: done

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 15 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: removed

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 14 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: done

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @dpoeschl  This test is failing intermittently: CSharpCompletionCommandHandlerTests_Projections.TestInObjectCreationExpression.

Could this be due to the recent changes going on around completion command handling?

(Note: this is blocking a community PR).

**Comment by Neme12** [comment]
retest windows_coreclr_debug_prtest please

**Comment by Neme12** [comment]
retest windows_debug_spanish_unit32_prtest please

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 20]
```diff
@@ -2,17 +2,66 @@
...
     internal static class InitializeParameterHelpers
     {
-        public static SyntaxNode GetBody(BaseMethodDeclarationSyntax containingMember)
-            => containingMember.Body ?? (SyntaxNode)containingMember.ExpressionBody;
+        public static bool IsFunctionDeclaration(SyntaxNode node)
+            => node is BaseMethodDeclarationSyntax
+            || node is LocalFunctionStatementSyntax
+            || node is AnonymousFunctionExpressionSyntax;
+
+        public static IBlockOperation GetBlockOperation(SyntaxNode functionDeclaration, SemanticModel semanticModel, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: 💭 It feels like some of the helpers could be unified, but it wasn't obvious how it would work so this seems fine for now.

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 21]
```diff
@@ -462,6 +462,150 @@ public C(string s)
     Option(CSharpCodeStyleOptions.PreferExpressionBodiedConstructors, CSharpCodeStyleOptions.WhenPossibleWithSuggestionEnforcement)));
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInitializeParameter)]
```
[the remark this answers is not in the document]
Comment (Neme12) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: added in d5dbed4a1922671b935aec0d6b80a462760f07d8

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 22]
```diff
@@ -141,6 +137,14 @@ private bool IsIfNullCheck(IOperation statement, IParameterSymbol parameter)
                         return true;
                     }
                 }
+                else if (condition is IIsPatternOperation isPatternOperation &&
+                         isPatternOperation.Pattern is IConstantPatternOperation constantPattern)
+                {
+                    if (IsNullCheck(constantPattern.Value, isPatternOperation.Value, parameter))
```
[the remark this answers is not in the document]
Comment (Neme12) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: added in 9876619bf03ef565d93155fb5a4236e0ceae3fd2

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs (hunk 20 above)
Comment (Neme12) [reply src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs]: I tried a few approaches but nothing actually made things simpler overall. I don't think there's a way to get a block operation from a method-level declaration, local function or anonymous function using the same API, is there? Unfortunately I think this has to implemented separately and language-specific.

**Comment by jinujoseph** [comment]
@sharwell  any thing pending here ?   LGTM approved to merge for 15.8.preview3

**Comment by Neme12** [comment]
I think this is my first PR! 🎉 Thanks

**Comment by Neme12** [comment]
I think this is my first PR 🎉 Thanks!
